File: src/class018/C_BinaryTreeTraversalIteration.py
class TreeNode:  
    def __init__(self, v):  
        self.val = v  
        self.left = None  
        self.right = None  
  
def pre_order(head):  
    """先序遍历（非递归实现）"""  
    if head:  
        stack = [head]  
        while stack:  
            node = stack.pop()  
            print(node.val, end=' ')  
            if node.right:  
                stack.append(node.right)  
            if node.left:  
                stack.append(node.left)  
        print()  
  
def in_order(head):  
    """中序遍历（非递归实现）"""  
    stack = []  
    while stack or head:  
        if head:  
            stack.append(head)  
            head = head.left  
        else:  
            head = stack.pop()  
            print(head.val, end=' ')  
            head = head.right  
    print()  
  
def pos_order_two_stacks(head):  
    """后序遍历，使用两个栈（非递归实现）"""  
    if head:  
        stack1 = [head]  
        stack2 = []  
        while stack1:  
            node = stack1.pop()  
            stack2.append(node)  
            if node.left:  
                stack1.append(node.left)  
            if node.right:  
                stack1.append(node.right)  
        while stack2:  
            node = stack2.pop()  
            print(node.val, end=' ')  
        print()  
  

# 后序打印所有节点，非递归版
# 这是用一个栈的方法
def pos_order_one_stack(h):
    if h is not None:
        stack = []
        stack.append(h)
        # 如果始终没有打印过节点，h就一直是头节点
        # 一旦打印过节点，h就变成打印节点
        # 之后h的含义 : 上一次打印的节点
        while stack:
            cur = stack[-1] # stack.peek()
            if cur.left is not None and h != cur.left and h != cur.right:
                # 有左树且左树没处理过
                stack.append(cur.left)
            elif cur.right is not None and h != cur.right:
                # 有右树且右树没处理过
                stack.append(cur.right)
            else:
                # 左树、右树 没有 或者 都处理过了
                print(cur.val, end=" ")
                h = stack.pop()
        print()
  
# 创建二叉树  
root = TreeNode(1)  
root.left = TreeNode(2)  
root.right = TreeNode(3)  
root.left.left = TreeNode(4)  
root.left.right = TreeNode(5)  
root.right.left = TreeNode(6)  
root.right.right = TreeNode(7)  
  
# 执行遍历  
pre_order(root)  
print("先序遍历非递归版")  
in_order(root)  
print("中序遍历非递归版")  
pos_order_two_stacks(root)  
print("后序遍历非递归版 - 2个栈实现")  
pos_order_one_stack(root)  
print("后序遍历非递归版 - 1个栈实现")  

File: src/class018/C_BinaryTreeTraversalIteration.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

// 定义二叉树节点结构
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

// 先序遍历（非递归实现）
void preOrder(TreeNode *head) {
    if (head != nullptr) {
        stack<TreeNode*> s;
        s.push(head);
        while (!s.empty()) {
            head = s.top(); s.pop();
            cout << head->val << " ";
            if (head->right != nullptr) {
                s.push(head->right);
            }
            if (head->left != nullptr) {
                s.push(head->left);
            }
        }
        cout << endl;
    }
}

// 中序遍历（非递归实现）
void inOrder(TreeNode *head) {
    stack<TreeNode*> s;
    while (!s.empty() || head != nullptr) {
        if (head != nullptr) {
            s.push(head);
            head = head->left;
        } else {
            head = s.top(); s.pop();
            cout << head->val << " ";
            head = head->right;
        }
    }
    cout << endl;
}

// 后序遍历，使用两个栈（非递归实现）
void posOrderTwoStacks(TreeNode *head) {
    if (head != nullptr) {
        stack<TreeNode*> s1, s2;
        s1.push(head);
        while (!s1.empty()) {
            head = s1.top(); s1.pop();
            s2.push(head);
            if (head->left != nullptr) {
                s1.push(head->left);
            }
            if (head->right != nullptr) {
                s1.push(head->right);
            }
        }
        while (!s2.empty()) {
            head = s2.top(); s2.pop();
            cout << head->val << " ";
        }
        cout << endl;
    }
}

// 后序遍历，使用一个栈（非递归实现）
void posOrderOneStack(TreeNode *h) {
    if (h != nullptr) {
        stack<TreeNode*> s;
        TreeNode *lastPrint = nullptr;
        s.push(h);
        while (!s.empty()) {
            TreeNode *cur = s.top();
            if (cur->left != nullptr && lastPrint != cur->left && lastPrint != cur->right) {
                s.push(cur->left);
            } else if (cur->right != nullptr && lastPrint != cur->right) {
                s.push(cur->right);
            } else { 
                cout << cur->val << " ";
                lastPrint = cur;
                s.pop();
            }
        }
        cout << endl;
    }
}

// 测试主函数
int main() {
    TreeNode *head = new TreeNode(1);
    head->left = new TreeNode(2);
    head->right = new TreeNode(3);
    head->left->left = new TreeNode(4);
    head->left->right = new TreeNode(5);
    head->right->left = new TreeNode(6);
    head->right->right = new TreeNode(7);

    preOrder(head);
    cout << "先序遍历非递归版" << endl;
    inOrder(head);
    cout << "中序遍历非递归版" << endl;
    posOrderTwoStacks(head);
    cout << "后序遍历非递归版 - 2个栈实现" << endl;
    posOrderOneStack(head);
    cout << "后序遍历非递归版 - 1个栈实现" << endl;

    return 0;
}
